### PR TITLE
Schedule tiller pod to desired node using node affinity

### PR DESCRIPTION
**Pull request by stoader**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | no
| API breaks?     | no
| Deprecations?   | no
| License         | Apache 2.0

### What's in this PR?

Use node affinity instead of node selector to schedule tiller pod master or system node where available as node affinity allows soft placement rule.

### Why?

This allows for scheduling tiller for a non-worker node to ensure that tiller is scheduled to a node that has enough resources. 

### Checklist

- [x] Implementation tested (with at least one cloud provider)
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/pipeline/blob/master/docs/error-handling-guide.md)
- [x] Logging code meets the guideline (TODO)

### CODE REVIEW

File: cluster/hooks.go [hunk 1]
```diff
@@ -618,15 +619,53 @@ func InstallHelmPostHook(cluster CommonCluster) error {
...
+								Operator: v1.NodeSelectorOpExists,
+							},
+						},
+					},
+				},
+			},
+		}
+	} else {
+		headNodePoolName := viper.GetString(pipConfig.PipelineHeadNodePoolName)
+		if len(headNodePoolName) > 0 {
```
Comment (siliconbrain) [inline cluster/hooks.go]:
```suggestion
		if headNodePoolName != "" {
```

File: cluster/hooks.go [hunk 2]
```diff
@@ -618,15 +619,53 @@ func InstallHelmPostHook(cluster CommonCluster) error {
...
+			if cluster.NodePoolExists(headNodePoolName) {
+				helmInstall.Tolerations = getHeadNodeTolerations()                   // add toleration for system node
+				helmInstall.NodeAffinity = getHeadNodeAffinity(cluster).NodeAffinity // try to schedule to system node
+			} else {
+				log.Warnf("head node pool %v not found, tiller deployment is not targeted to any node pool.", headNodePoolName)
```
Comment (siliconbrain) [inline cluster/hooks.go]:
```suggestion
				log.Warnf("head node pool %q not found, tiller deployment is not targeted to any node pool.", headNodePoolName)
```

File: helm/install.go [hunk 3]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
 		AutoMountServiceAccountToken: true,
 	}
 
-	if len(helmInstall.TargetNodePool) > 0 {
+	for i := range helmInstall.Tolerations {
+		if len(helmInstall.Tolerations[i].Key) > 0 {
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
		if helmInstall.Tolerations[i].Key != "" {
```

File: helm/install.go [hunk 4]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%v].key=%v", i, helmInstall.Tolerations[i].Key))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%d].key=%s", i, helmInstall.Tolerations[i].Key))
```

File: helm/install.go [hunk 5]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+		}
+
+		if len(helmInstall.Tolerations[i].Operator) > 0 {
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
		if helmInstall.Tolerations[i].Operator != "" {
```

File: helm/install.go [hunk 6]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%v].operator=%v", i, helmInstall.Tolerations[i].Operator))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%d].operator=%s", i, helmInstall.Tolerations[i].Operator))
```

File: helm/install.go [hunk 7]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+		}
+
+		if len(helmInstall.Tolerations[i].Value) > 0 {
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
		if helmInstall.Tolerations[i].Value != "" {
```

File: helm/install.go [hunk 8]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%v].value=%v", i, helmInstall.Tolerations[i].Value))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%d].value=%s", i, helmInstall.Tolerations[i].Value))
```

File: helm/install.go [hunk 9]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+		}
+
+		if len(helmInstall.Tolerations[i].Effect) > 0 {
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
		if helmInstall.Tolerations[i].Effect != "" {
```

File: helm/install.go [hunk 10]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%v].effect=%v", i, helmInstall.Tolerations[i].Effect))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			opts.Values = append(opts.Values, fmt.Sprintf("spec.template.spec.tolerations[%d].effect=%s", i, helmInstall.Tolerations[i].Effect))
```

File: helm/install.go [hunk 11]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+		}
+	}
+
+	if helmInstall.NodeAffinity != nil {
+		for i := range helmInstall.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
+			preferredSchedulingTerm := helmInstall.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution[i]
+
+			schedulingTermString := fmt.Sprintf("spec.template.spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[%v]", i)
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			schedulingTermString := fmt.Sprintf("spec.template.spec.affinity.nodeAffinity.preferredDuringSchedulingIgnoredDuringExecution[%d]", i)
```

File: helm/install.go [hunk 12]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+			opts.Values = append(opts.Values, fmt.Sprintf("%s.weight=%v", schedulingTermString, preferredSchedulingTerm.Weight))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
			opts.Values = append(opts.Values, fmt.Sprintf("%s.weight=%d", schedulingTermString, preferredSchedulingTerm.Weight))
```

File: helm/install.go [hunk 13]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+
+			for j := range preferredSchedulingTerm.Preference.MatchExpressions {
+				matchExpression := preferredSchedulingTerm.Preference.MatchExpressions[j]
+
+				matchExpressionString := fmt.Sprintf("%s.preference.matchExpressions[%v]", schedulingTermString, j)
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
				matchExpressionString := fmt.Sprintf("%s.preference.matchExpressions[%d]", schedulingTermString, j)
```

File: helm/install.go [hunk 14]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+
+				opts.Values = append(opts.Values, fmt.Sprintf("%s.key=%v", matchExpressionString, matchExpression.Key))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
				opts.Values = append(opts.Values, fmt.Sprintf("%s.key=%s", matchExpressionString, matchExpression.Key))
```

File: helm/install.go [hunk 15]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+				opts.Values = append(opts.Values, fmt.Sprintf("%s.operator=%v", matchExpressionString, matchExpression.Operator))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
				opts.Values = append(opts.Values, fmt.Sprintf("%s.operator=%s", matchExpressionString, matchExpression.Operator))
```

File: helm/install.go [hunk 16]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+
+				for k := range matchExpression.Values {
+					opts.Values = append(opts.Values, fmt.Sprintf("%s.values[%v]=%v", matchExpressionString, k, matchExpression.Values[i]))
```
Comment (siliconbrain) [inline helm/install.go]:
```suggestion
					opts.Values = append(opts.Values, fmt.Sprintf("%s.values[%d]=%v", matchExpressionString, k, matchExpression.Values[i]))
```

File: helm/install.go [hunk 17]
```diff
@@ -322,15 +321,55 @@ func Install(log logrus.FieldLogger, helmInstall *phelm.Install, kubeConfig []by
...
+	/*if len(helmInstall.TargetNodePool) > 0 {
 		opts.Values = []string{
 			fmt.Sprintf("spec.template.spec.tolerations[0].key=%v", pkgCommon.HeadNodeTaintKey),
 			"spec.template.spec.tolerations[0].operator=Equal",
 			fmt.Sprintf("spec.template.spec.tolerations[0].value=%v", helmInstall.TargetNodePool),
 		}
 		// TODO check why this even needed? Soft affinity would be better here.
 		//opts.NodeSelectors = fmt.Sprintf("%s=%s", pkgCommon.LabelKey, helmInstall.TargetNodePool)
-	}
+	}*/
```
Comment (orymate) [inline helm/install.go]: remove this block

**Pull request merged**